Source/WebKit/Shared/Cocoa/CoreIPCNSURLCredential.mm

```diff
@@ -26,9 +26,214 @@
+                vector.append(std::make_pair(WTFMove(k), WTFMove(*option)));
+            }
+        }
+        m_data.attributes = WTFMove(vector);
+    }
+
+    SET_MDATA(identifier, NSString, CoreIPCString);
+
+    NSNumber *useKeychain = dict[@"useKeychain"];
+    if ([useKeychain isKindOfClass:NSNumber.class])
+        m_data.useKeychain = [useKeychain boolValue];
+
+    SecTrustRef secTrust = static_cast<SecTrustRef>(dict[@"trust"]);
+    if (secTrust && CFGetTypeID(secTrust) == SecTrustGetTypeID())
+        m_data.trust = CoreIPCSecTrust(secTrust);
+    else
+        m_data.trust = CoreIPCSecTrust(nil);
+
+    SET_MDATA(service, NSString, CoreIPCString);
+
+    NSDictionary *flags = dict[@"flags"];
+    if ([flags isKindOfClass:NSDictionary.class]) {
+        Vector<WebKit::CoreIPCNSURLCredentialData::Flags> vector;
+        vector.reserveCapacity(flags.count);
+        for (NSString *key in attributes) {
+            if (![key isKindOfClass:NSString.class]) {
+                ASSERT_NOT_REACHED();
+                break;
+            }
+            NSString *value = [flags objectForKey:key];
+            if (![value isKindOfClass:NSString.class]) {
+                ASSERT_NOT_REACHED();
+                break;
+            }
+            auto k = CoreIPCString(key);
+            auto v = CoreIPCString(value);
+            vector.append(std::make_pair(WTFMove(k), WTFMove(v)));
+        }
+        m_data.flags = WTFMove(vector);
+    }
+
+    SET_MDATA(uuid, NSString, CoreIPCString);
+    SET_MDATA(appleID, NSString, CoreIPCString);
+    SET_MDATA(realm, NSString, CoreIPCString);
+    SET_MDATA(token, NSString, CoreIPCString);
+}
+
+CoreIPCNSURLCredential::CoreIPCNSURLCredential(CoreIPCNSURLCredentialData&& data)
+: m_data(WTFMove(data)) { }
+
+RetainPtr<id> CoreIPCNSURLCredential::toID() const
+{
+    auto dict = adoptNS([[NSMutableDictionary alloc] initWithCapacity:7]);
+
+    RetainPtr<NSNumber> persistence;
+    switch (m_data.persistence) {
+    case CoreIPCNSURLCredentialPersistence::None:
+        persistence = adoptNS([NSNumber numberWithInt:kCFURLCredentialPersistenceNone]);
+        break;
+    case CoreIPCNSURLCredentialPersistence::Session:
+        persistence = adoptNS([NSNumber numberWithInt:kCFURLCredentialPersistenceForSession]);
+        break;
+    case CoreIPCNSURLCredentialPersistence::Permanent:
+        persistence = adoptNS([NSNumber numberWithInt:kCFURLCredentialPersistencePermanent]);
+        break;
+    case CoreIPCNSURLCredentialPersistence::Synchronizable:
+        persistence = adoptNS([NSNumber numberWithInt:kCFURLCredentialPersistenceSynchronizable]);
+        break;
+    default:
+        ASSERT_NOT_REACHED();
+        persistence = adoptNS([NSNumber numberWithInt:kCFURLCredentialPersistenceNone]);
+        break;
+    }
+    [dict setObject:persistence.get() forKey:@"persistence"];
+
+    if (m_data.user && m_data.password) {
+        [dict setObject:@(kURLCredentialInternetPassword) forKey:@"type"];
```

Comment:
It's a bit weird that we decide `type` value based on what members are available. Can't we just store the type in `CoreIPCNSURLCredentialData`? Then here we could check corresponding members.
E.g. 
```
if (m_data.type == kURLCredentialInternetPassword) {
    // Assert  m_data.user and m_data.password is available
    // Construct dict as needed 
}
```

Reply:
The members which are set let us know which type it is. We could switch on a value in the struct but that is coming from an attacker controlled WebContent Process.

Reply:
hmm the other parameters in `CoreIPCNSURLCredentialData` are also extracted from IPC message, right? Why can we trust the other paramters but not type?

Reply:
For deserialization we cannot trust any of the values. We could switch on an enum and then validate or we can infer it by the items in the struct we have received.

Reply:
(For NSURLCredential the flow would be NP -> UIP, not WCP)

Reply:
> or we can infer it by the items in the struct we have received.

I don't think we should infer it when it can be explicitly specified by sender. I noticed the weirdness because the existing implementation actually resets "type" and "realm" properties at `if (m_data.appleID && m_data.realm && m_data.password)` and `if (m_data.token && m_data.realm)`. 

If the sender process sets all these properties (no matter it is under attack or there is a bug), we won't get the type right. We could fix the issue with `else if (m_data.token && m_data.realm)`, but it still seems fragile since the process could still set all properties and the type result will be based on which type the code checks first.

So I think a more robust way is to handle properties based on the specified type, and we should only handle it once (if the properties do not match specified type, we just fail instead of trying to match next type).